File: ios/FoundationApp/CodespacesPreview.swift
import Foundation

enum CodespacesPreview {
    static func request(_ request: URLRequest, rootURL: URL) -> URLRequest {
        guard
            rootURL.scheme == "https",
            let host = rootURL.host?.lowercased(),
            host.range(of: #"^[a-z0-9-]+-[0-9]+\.app\.github\.dev$"#, options: .regularExpression) != nil,
            let url = request.url,
            url.scheme == "https",
            url.host?.lowercased() == host,
            (url.port ?? 443) == (rootURL.port ?? 443),
            url.user == nil,
            url.password == nil
        else {
            return request
        }

        var previewRequest = request
        // Codespaces' HTML warning is not a Turbo page. This skips that warning,
        // while the tunnel's public/private access control still applies.
        previewRequest.setValue("true", forHTTPHeaderField: "X-Tunnel-Skip-AntiPhishing-Page")
        return previewRequest
    }
}

File: android/app/src/main/java/com/firstdraft/foundation/CodespacesPreview.kt
package com.firstdraft.foundation

import java.net.URI
import java.net.URISyntaxException

object CodespacesPreview {
    private val codespaceHost = Regex("[a-z0-9-]+-[0-9]+\\.app\\.github\\.dev")

    fun headers(url: String, root: String, production: String, debug: Boolean): Map<String, String> {
        if (!debug || root == production) return emptyMap()
        val origin = uri(root) ?: return emptyMap()
        val request = uri(url) ?: return emptyMap()
        val matching = origin.scheme == "https" && request.scheme == "https" &&
            origin.host != null && codespaceHost.matches(origin.host) &&
            request.host == origin.host && port(request) == port(origin) &&
            origin.userInfo == null && request.userInfo == null
        return if (matching) mapOf("X-Tunnel-Skip-AntiPhishing-Page" to "true") else emptyMap()
    }

    private fun port(uri: URI): Int = if (uri.port == -1) 443 else uri.port

    private fun uri(value: String): URI? = try {
        URI(value)
    } catch (_: URISyntaxException) {
        null
    }
}
